### PR TITLE
feat: suport read/write Manifest

### PR DESCRIPTION
**Pull request by ZENOTME**
This PR prepare to support read/write Manifest. 

related issue: #36

### CODE REVIEW

**Comment by ZENOTME** [comment]
For now, it still not been completed. 

It only complete the basic design and I want to make sure whether the design well first. 

If it looks well, I will complete it and add the test later.

cc @JanKaul @Fokko @Xuanwo @liurenjie1024

File: crates/iceberg/src/spec/partition.rs [hunk 1]
```diff
@@ -54,6 +56,24 @@ impl PartitionSpec {
     pub fn builder() -> PartitionSpecBuilder {
         PartitionSpecBuilder::default()
     }
+
+    /// Construct the partition type from schema
+    pub fn partition_type(&self, schema: &Schema) -> Result<StructType, Error> {
+        let mut fields = Vec::with_capacity(self.fields.len());
+        for field in &self.fields {
+            let field = schema.field_by_id(field.source_id).ok_or_else(|| {
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/partition.rs:64]: I think this is incorrect? The partition type should be determined by type after transformation.

File: crates/iceberg/src/spec/values.rs [hunk 2]
```diff
@@ -966,6 +978,547 @@ mod timestamptz {
     }
 }
 
+mod serde {
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/values.rs:981]: I would suggest to move this into another pr and resolve them first, otherwise this pr would get too large to review.

**Review by liurenjie1024 — COMMENTED** [review]
@ZENOTME Thanks for this! I would suggest to move the part for ser/de to another pr and resolve them first, otherwise it would be too difficult to review.

I'm ok with the overall design.

File: crates/iceberg/src/spec/values.rs (hunk 2 above)
Comment (ZENOTME) [reply crates/iceberg/src/spec/values.rs:981]: If `RawLiteral` looks well, I can seperate it to another PR.

File: crates/iceberg/src/spec/partition.rs (hunk 1 above)
Comment (ZENOTME) [reply crates/iceberg/src/spec/partition.rs:64]: Yes. My bad.🥵

File: crates/iceberg/src/spec/values.rs (hunk 2 above)
Comment (liurenjie1024) [reply crates/iceberg/src/spec/values.rs:981]: I'm ok with this approach, cc @JanKaul @Xuanwo

Comment (JanKaul) [reply crates/iceberg/src/spec/values.rs:981]: I think using the RawLiteral makes sense. Initially I tried to make `Literal` directly deserializable from avro but the types have to be converted from avro types to iceberg types anyway.

File: crates/iceberg/src/spec/manifest.rs [hunk 3]
```diff
@@ -0,0 +1,671 @@
...
+                sequence_number: None,
+                file_sequence_number: None,
+                data_file: self.data_file.try_into(&partition_type, schema)?,
+            })
+        }
+    }
+
+    #[serde_as]
+    #[derive(Serialize, Deserialize)]
+    pub(super) struct DataFile {
```
Comment (JanKaul) [inline crates/iceberg/src/spec/manifest.rs:513]: Just to be sure. Are the `block_size_in_bytes`, `file_ordinal`, `sort_columns` fields also deprecated in the v1 `Datafile`?

File: crates/iceberg/src/spec/manifest.rs [hunk 4]
```diff
@@ -0,0 +1,671 @@
...
+            FormatVersion::V2 => {
+                let reader = AvroReader::with_schema(Self::v2_schema(), bs)?;
+                for value in reader {
+                    entries.push(
+                        from_value::<_serde::ManifestEntryV2>(&value?)?
+                            .try_into(&partition_type, &metadata.schema)?,
+                    );
+                }
+            }
+        };
```
Comment (JanKaul) [inline crates/iceberg/src/spec/manifest.rs:65]: Would it be possible to iterate over the `reader` values, map the values to `ManifestEntry` and then collect the iterator combinator into the `entries` vec. This way you might save on reallocations of the `entries` vector.

File: crates/iceberg/src/spec/manifest.rs (hunk 3 above)
Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest.rs:513]: [According to pyiceberg](https://github.com/apache/iceberg-python/blob/195d4aaf23f64e853e218305688690d7f3658c14/pyiceberg/manifest.py#L126), seems remain `block_size_in_bytes` in v1 but other fields deprecated. cc @Fokko

File: crates/iceberg/src/spec/manifest.rs [hunk 5]
```diff
@@ -0,0 +1,671 @@
...
+        }
+    }
+
+    #[serde_as]
+    #[derive(Serialize, Deserialize)]
+    #[cfg_attr(test, derive(Debug, PartialEq, Eq))]
+    struct BytesEntry {
+        key: i32,
+        #[serde_as(as = "Bytes")]
+        value: Vec<u8>,
```
Comment (JanKaul) [inline crates/iceberg/src/spec/manifest.rs:612]: Would it make sense to use `serde_bytes::ByteBuf` here? It is supposed to have an optimized serialization/deserialization.

File: crates/iceberg/src/spec/manifest.rs (hunk 4 above)
Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest.rs:65]: Yes!

File: crates/iceberg/src/spec/manifest.rs [hunk 6]
```diff
@@ -0,0 +1,671 @@
...
+    /// ID of the schema used to write the manifest as a string
+    schema_id: i32,
+    /// The partition spec used  to write the manifest
+    partition_spec: PartitionSpec,
+    /// ID of the partition spec used to write the manifest as a string
+    partition_spec_id: i32,
+    /// Table format version number of the manifest as a string
+    format_version: FormatVersion,
+    /// Type of content files tracked by the manifest: “data” or “deletes”
+    content: ManifestContentType,
```
Comment (Fokko) [inline crates/iceberg/src/spec/manifest.rs:94]: This one should be omitted for V1 (or just set it to `data` as a default).

Comment (Fokko) [reply crates/iceberg/src/spec/manifest.rs:94]: Ah, this is done below 👍

File: crates/iceberg/src/spec/manifest.rs (hunk 3 above)
Comment (Fokko) [reply crates/iceberg/src/spec/manifest.rs:513]: Yes, I would leave all three of them out of the struct. Only when writing do you need to make sure that you include `block_size_in_bytes`. You can do this very cleverly using the Icebergs' `write-default`.

File: crates/iceberg/src/spec/manifest.rs [hunk 7]
```diff
@@ -0,0 +1,671 @@
...
+    #[serde_as]
+    #[derive(Serialize, Deserialize)]
+    #[cfg_attr(test, derive(Debug, PartialEq, Eq))]
+    struct BytesEntry {
+        key: i32,
+        #[serde_as(as = "Bytes")]
+        value: Vec<u8>,
+    }
+
+    fn parse_bytes_entry(
```
Comment (Fokko) [inline crates/iceberg/src/spec/manifest.rs:615]: Nice one!

File: crates/iceberg/src/spec/values.rs [hunk 8]
```diff
@@ -966,6 +978,547 @@ mod timestamptz {
...
+        de::Visitor,
+        ser::{SerializeSeq, SerializeStruct},
+        Deserialize, Serialize,
+    };
+    use serde_derive::Deserialize as DeserializeDerive;
+    use serde_derive::Serialize as SerializeDerive;
+
+    #[derive(SerializeDerive, DeserializeDerive)]
+    #[serde(transparent)]
+    /// Raw literal representation used for serde. The seriailize way is used for avro serializer.
```
Comment (Fokko) [inline crates/iceberg/src/spec/values.rs:1000]:
```suggestion
    /// Raw literal representation used for serde. The serialize way is used for Avro serializer.
```

**Comment by ZENOTME** [comment]
I will split the RawLiteral out into another PR.

**Comment by ZENOTME** [comment]
I have send a new PR #82 for value part. Let's complete it first and then continue on working this PR later.

**Comment by ZENOTME** [comment]
Seem the binary file will make typos fail. Maybe we should fix #70 first. Other parts is ready to review.

File: crates/iceberg/src/spec/manifest.rs [hunk 9]
```diff
@@ -0,0 +1,1847 @@
...
+    /// Total file size in bytes
+    file_size_in_bytes: i64,
+    /// field id: 108
+    /// key field id: 117
+    /// value field id: 118
+    ///
+    /// Map from column id to the total size on disk of all regions that
+    /// store the column. Does not include bytes necessary to read other
+    /// columns, like footers. Leave null for row-oriented formats (Avro)
+    column_sizes: Option<HashMap<i32, i64>>,
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/manifest.rs:921]: Using `HashMap` directly? This will simplify operators on these fields. Same as belowing fields.

**Review by liurenjie1024 — COMMENTED** [review]
This PR looks great! We are close to manipulating metadata in rust! Left some small comments to improve.

File: crates/iceberg/src/spec/manifest_list.rs [hunk 10]
```diff
@@ -30,6 +30,9 @@ use self::{
 
 use super::{FormatVersion, StructType};
 
+/// The seq number when no added files are present.
+pub const UNASSIGNED_SEQ_NUMBER: i64 = -1;
```
[the remark this answers is not in the document]
Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest_list.rs:34]: It's for [min_sequence_number](https://iceberg.apache.org/spec/#manifests:~:text=516-,min_sequence_number,-long). Which means that all manifest entry is deleted. Actually, I am not sure whether it's the correct process way because there is not `UNASSIGNED_SEQ_NUMBER` mentioned in spec.🤔 Does pyiceberg use `UNASSIGNED_SEQ_NUMBER`? cc @Fokko

File: crates/iceberg/src/spec/transform.rs [hunk 11]
```diff
@@ -126,17 +126,20 @@ pub enum Transform {
 impl Transform {
     /// Get the return type of transform given the input type.
     /// Returns `None` if it can't be transformed.
-    pub fn result_type(&self, input_type: &Type) -> Option<Type> {
+    pub fn result_type(&self, input_type: &Type) -> Result<Type> {
```
[the remark this answers is not in the document]
Comment (ZENOTME) [reply crates/iceberg/src/spec/transform.rs:129]: The None case is the case which is not support in iceberg.🤔 Why not treat it as a error.

File: crates/iceberg/src/spec/manifest.rs [hunk 12]
```diff
@@ -0,0 +1,1847 @@
...
+                let field_summary = self.field_summary.entry(k).or_default();
+                if v > 0 {
+                    field_summary.contains_null = true;
+                }
+            }
+        }
+        if let Some(nan) = &entry.data_file.nan_value_counts {
+            for (&k, &v) in nan {
+                let field_summary = self.field_summary.entry(k).or_default();
+                assert!(v >= 0);
```
[the remark this answers is not in the document]
Comment (Xuanwo) [reply crates/iceberg/src/spec/manifest.rs:142]: Looks good to me.

File: crates/iceberg/src/spec/manifest_list.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (liurenjie1024) [reply crates/iceberg/src/spec/manifest_list.rs:34]: I see. Maybe we don't need this, and using `Option` is more appropriate?

File: crates/iceberg/src/spec/transform.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (liurenjie1024) [reply crates/iceberg/src/spec/transform.rs:129]: They have different meaning. When return `Option`, it acts more like a check to say if this transform can be applied, and whether it's an error is up to user. But their difference seems negligible. cc @Xuanwo @JanKaul @Fokko What do you think?

File: crates/iceberg/src/spec/manifest_list.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Fokko) [reply crates/iceberg/src/spec/manifest_list.rs]: In PyIceberg we set it to `0` when it is not set (v2) or unknown (v1). It is used to effectively prune and delete files that are not relevant to the data that are being read.

In PyIceberg we first do the normal query planning by applying the partition filtering and the metrics. The new end up with a list of files where we compute the minimal data file sequence number:

https://github.com/apache/iceberg-python/blob/8c8abb5c4c258e32941110a9ce0938e1328290b3/pyiceberg/table/__init__.py#L1028-L1037

There is an obvious fallback to `INITIAL_SEQUENCE_NUMBER` which is `0`. If this happens then we know that we can't use this number to prune, and all the deletes files that are present will be included (because the sequence number there is also greater than or equal to zero.

I would suggest setting this to zero.

File: crates/iceberg/src/spec/transform.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Fokko) [reply crates/iceberg/src/spec/transform.rs:129]: I think there is a difference, let me explain! In Iceberg it is possible to have your custom transform, for example if you want to do something very specific, or you want to influence the distribution when it comes to partitioning. It can be that the transform is not available in the client, and then we cast it to an `UnknownTransform`: https://github.com/apache/iceberg-python/blob/8c8abb5c4c258e32941110a9ce0938e1328290b3/pyiceberg/transforms.py#L688

What this essentially means is that we don't know how it is transformed, and we cannot apply speedups. For example, `bucket[16](34) = 3` then we know in which bucket the data is that we're looking for. But if there is some custom partitioning then we just have to look through all the buckets.

To this specific case of the type, in PyIceberg we return a `StringType()`, that's borrowed from Java where you just cast it to a string. I don't think that's used anywhere since we don't call this on the `UnknownTransform`.

File: crates/iceberg/src/spec/manifest.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Fokko) [reply crates/iceberg/src/spec/manifest.rs]: There is no harm in adding the check, right?

[the remark this answers is not in the document]
Comment (liurenjie1024) [reply crates/iceberg/src/spec/manifest.rs]: If this check failed, e.g. some bug happens, the program will crash.

File: crates/iceberg/src/spec/manifest_list.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (liurenjie1024) [reply crates/iceberg/src/spec/manifest_list.rs]: LGTM.

File: crates/iceberg/src/spec/manifest.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest.rs]: > There is no harm in adding the check, right?

Yes, but we can't guarantee add check in every place, it easy to miss the check. So use u64 in memory struct can guarantee that the value is always > 0. We only need to add the check when convert it into serde to write.

File: crates/iceberg/src/spec/transform.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (liurenjie1024) [reply crates/iceberg/src/spec/transform.rs:129]: LGTM.

**Comment by liurenjie1024** [comment]
I think we need to add configurations in `typos.toml` to skip the check:

https://github.com/crate-ci/typos/blob/master/docs/reference.md

File: crates/iceberg/src/spec/manifest_list.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest_list.rs]: I guess for v2, here the min sequence number should be set as sequence number.
https://github.com/apache/iceberg-python/blob/8c8abb5c4c258e32941110a9ce0938e1328290b3/pyiceberg/manifest.py#L934C1-L934C1

> The minimum data sequence number of all live data or delete files in the manifest; use 0 when reading v1 manifest lists

I find that the minimum data sequence will be set if it the manifest has live files(the file exist before), otherwise it will be set as sequence number in manifest list writer.
https://github.com/apache/iceberg-python/blob/8c8abb5c4c258e32941110a9ce0938e1328290b3/pyiceberg/manifest.py#L733C1-L733C1

And I realize that I should not put the sequence number in manifest writer. The sequence number only need set in manifest list. For manifest entry, [the sequence number should be in read process](https://iceberg.apache.org/spec/#sequence-number-inheritance).

**Comment by Xuanwo** [comment]
> I think we need to add configurations in `typos.toml` to skip the check:
> 
> https://github.com/crate-ci/typos/blob/master/docs/reference.md

Yep, we should skip check on `testdata`.

**Comment by ZENOTME** [comment]
> > I think we need to add configurations in `typos.toml` to skip the check:
> > https://github.com/crate-ci/typos/blob/master/docs/reference.md
> 
> Yep, we should skip check on `testdata`.

I have added `.typo.toml` to skip.

File: crates/iceberg/src/spec/manifest_list.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest_list.rs]: I have add a commit to fix it. PTAL @liurenjie1024  @Fokko

File: crates/iceberg/src/spec/partition.rs [hunk 13]
```diff
@@ -69,6 +71,30 @@ impl PartitionSpec {
                 .iter()
                 .all(|f| matches!(f.transform, Transform::Void))
     }
+
+    /// Returns the partition type of this partition spec.
+    pub fn partition_type(&self, schema: &Schema) -> Result<StructType, Error> {
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/partition.rs:76]: Could we add some ut for this? What should be the return type if no partition field?

File: crates/iceberg/src/spec/manifest_list.rs [hunk 14]
```diff
@@ -156,12 +171,36 @@ impl ManifestListWriter {
...
+                        if manifest_entry.added_snapshot_id != self.snapshot_id {
+                            return Err(Error::new(
+                                ErrorKind::DataInvalid,
+                                format!(
+                                    "Found unassigned sequence number for a manifest from snapshot {}.",
+                                    manifest_entry.added_snapshot_id
+                                ),
+                            ));
+                        }
+                        manifest_entry.sequence_number = self.sequence_number;
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/manifest_list.rs:190]: This assignment is meaningless to me. This manifest entry will not return to user.

File: crates/iceberg/src/spec/manifest_list.rs [hunk 15]
```diff
@@ -156,12 +171,36 @@ impl ManifestListWriter {
...
                 for manifest_entry in manifest_entries {
-                    let manifest_entry: ManifestListEntryV1 = manifest_entry.into();
+                    let manifest_entry: ManifestListEntryV1 = manifest_entry.try_into()?;
                     self.avro_writer.append_ser(manifest_entry)?;
                 }
             }
             FormatVersion::V2 => {
-                for manifest_entry in manifest_entries {
+                for mut manifest_entry in manifest_entries {
+                    if manifest_entry.sequence_number == UNASSIGNED_SEQUENCE_NUMBER {
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/manifest_list.rs:180]: It seems weird to me to check this in a writer. I think this is supposed to do in transaction api.

File: crates/iceberg/src/spec/manifest.rs [hunk 16]
```diff
@@ -0,0 +1,1847 @@
...
+    /// Total file size in bytes
+    file_size_in_bytes: u64,
+    /// field id: 108
+    /// key field id: 117
+    /// value field id: 118
+    ///
+    /// Map from column id to the total size on disk of all regions that
+    /// store the column. Does not include bytes necessary to read other
+    /// columns, like footers. Leave null for row-oriented formats (Avro)
+    column_sizes: Option<HashMap<i32, u64>>,
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/manifest.rs:917]: Use `HashMap` directly here? Same as below.

File: crates/iceberg/src/spec/manifest.rs [hunk 17]
```diff
@@ -0,0 +1,1847 @@
...
+    added_files: u32,
+    added_rows: u64,
+    existing_files: u32,
+    existing_rows: u64,
+    deleted_files: u32,
+    deleted_rows: u64,
+
+    min_seq_num: Option<i64>,
+
+    key_metadata: Option<Vec<u8>>,
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/manifest.rs:99]: Use `Vec` directly?

File: crates/iceberg/src/spec/manifest.rs [hunk 18]
```diff
@@ -0,0 +1,1847 @@
...
+        Lazy::new(|| {
+            Arc::new(NestedField::optional(
+                140,
+                "sort_order_id",
+                Type::Primitive(PrimitiveType::Int),
+            ))
+        })
+    };
+
+    pub fn manifest_schema_v2(partition_type: StructType) -> Result<AvroSchema, Error> {
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/manifest.rs:633]:
```suggestion
    pub(super) fn manifest_schema_v2(partition_type: StructType) -> Result<AvroSchema, Error> {
```

File: .typos.toml [hunk 19]
```diff
@@ -0,0 +1,2 @@
+[files]
+extend-exclude = ["crates/iceberg/testdata"]
```
Comment (liurenjie1024) [inline .typos.toml:2]: Is it possible to sth like `**/testdata`? I think all `testdata` should be skipped.

File: crates/iceberg/src/spec/manifest_list.rs (hunk 14 above)
Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest_list.rs:190]: The reason we postpone this assignment is that we don't need to recall manifest file if commit is fail. E.g. if the we assign the sequence_number in ManifestWriter
```
// write Manifest
let entry = ManifestWriter(seq_num);
// writer ManifestList
let res = ManifestListWriter(entry);
// if commit is failed, we should recall the ManifestWriter to get a new entry using a new seq num.
let res = commit(res);
```

File: crates/iceberg/src/spec/manifest_list.rs (hunk 15 above)
Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest_list.rs:180]: Yes, we also can do this check outside the writer.🤔 Looks reasonable.

File: crates/iceberg/src/spec/manifest_list.rs (hunk 14 above)
Comment (liurenjie1024) [reply crates/iceberg/src/spec/manifest_list.rs:190]: IIUC, the sequence number is optional and inherited from snapshot, when we do retry, we just need to retry adding snapshot.

Comment (liurenjie1024) [reply crates/iceberg/src/spec/manifest_list.rs:190]: BTW, could change the variable name to `manifest_list_entry`? Otherwise it's really confusing to read this🤣

File: crates/iceberg/src/spec/manifest_list.rs (hunk 15 above)
Comment (liurenjie1024) [reply crates/iceberg/src/spec/manifest_list.rs:180]: Sorry, let's resolve this, I confuse this with manifest

File: crates/iceberg/src/spec/partition.rs (hunk 13 above)
Comment (ZENOTME) [reply crates/iceberg/src/spec/partition.rs:76]: > What should be the return type if no partition field?

For now, it return a empty struct type

**Review by liurenjie1024 — APPROVED** [review]
LGTM, thanks for the effort!

File: crates/iceberg/src/spec/manifest.rs [hunk 20]
```diff
@@ -0,0 +1,1864 @@
...
+                    )),
+                    value_field: Arc::new(NestedField::required(
+                        124,
+                        "value",
+                        Type::Primitive(PrimitiveType::Long),
+                    )),
+                }),
+            ))
+        })
+    };
```
Comment (Fokko) [inline crates/iceberg/src/spec/manifest.rs:532]: Both Java and PyIceberg don't write `distinct_counts`

Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest.rs:532]: Seems the spec doesn't have info about this. Is that a consensus to deprecatei?🤔

File: crates/iceberg/src/spec/manifest_list.rs [hunk 21]
```diff
@@ -1120,15 +1236,17 @@ mod test {
...
+        let mut expected_manifest_list = ManifestList {
             entries: vec![ManifestListEntry {
                 manifest_path: "s3a://icebergdata/demo/s1/t1/metadata/05ffe08b-810f-49b3-a8f4-e88fc99b254a-m0.avro".to_string(),
                 manifest_length: 6926,
                 partition_spec_id: 1,
                 content: ManifestContentType::Data,
-                sequence_number: 1,
-                min_sequence_number: 1,
-                added_snapshot_id: 377075049360453639,
+                sequence_number: UNASSIGNED_SEQUENCE_NUMBER,
```
Comment (Fokko) [inline crates/iceberg/src/spec/manifest_list.rs:1247]: On the manifest list the snapshot should be assigned, maybe we can add a note here that it should be fixed once we get the snapshots generation in there.

File: crates/iceberg/src/spec/manifest.rs [hunk 22]
```diff
@@ -0,0 +1,1821 @@
...
+        });
+        assert!(manifest.metadata.partition_spec.fields.is_empty());
+        assert!(manifest.metadata.content == ManifestContentType::Data);
+        assert!(manifest.metadata.format_version == FormatVersion::V1);
+        assert_eq!(manifest.entries.len(), 4);
+        let entry = &manifest.entries[0];
+        assert!(entry.status == ManifestStatus::Added);
+        assert!(entry.snapshot_id == Some(2966623707104393227));
+        assert!(entry.sequence_number.is_none());
+        assert!(entry.file_sequence_number.is_none());
```
Comment (Fokko) [inline crates/iceberg/src/spec/manifest.rs:1635]: image

These should default to `0`. You could implement this using the initial-value of Iceberg:

image


```suggestion
        assert!(entry.sequence_number == Some(0));
        assert!(entry.file_sequence_number == Some(0));
```

File: crates/iceberg/src/spec/manifest.rs [hunk 23]
```diff
@@ -0,0 +1,1821 @@
...
+        assert!(manifest.metadata.partition_spec.fields.is_empty());
+        assert!(manifest.metadata.content == ManifestContentType::Data);
+        assert!(manifest.metadata.format_version == FormatVersion::V2);
+        // test entries
+        assert!(manifest.entries.len() == 1);
+        let entry = &manifest.entries[0];
+        assert!(entry.status == ManifestStatus::Added);
+        assert!(entry.snapshot_id == Some(0));
+        assert!(entry.sequence_number == Some(1));
+        assert!(entry.file_sequence_number == Some(1));
```
Comment (Fokko) [inline crates/iceberg/src/spec/manifest.rs:1397]: This is some interesting test-data 🤔 The status is `added`, so this means a new file which typically doesn't set the snapshot-id, sequence-number, and file-sequence-number since they should be inherited. This way we don't have to rewrite the manifest after a failed conflict (because a new snapshot-id and sequence-number is assigned).

File: crates/iceberg/src/spec/manifest.rs (hunk 20 above)
Comment (Fokko) [reply crates/iceberg/src/spec/manifest.rs]: Yes, it was never implemented in most engines since it is not cheap to compute this kind of statistics. The alternative now is to compute sketches and store them in Puffin files.

Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest.rs]: Got it! Thanks!

**Review by Fokko — APPROVED** [review]
LGTM! Thanks for putting so much effort into this @ZENOTME 🙌 Thanks for the review @JanKaul and @liurenjie1024. I think there are still some loose ends with the snapshot-id and sequence-number inheritance, but we can tackle that separately.

File: crates/iceberg/src/spec/manifest_list.rs (hunk 21 above)
Comment (ZENOTME) [reply crates/iceberg/src/spec/manifest_list.rs:1247]: I'm sorry. I'm not sure what you mean.🤔 In here the `snapshot id` is assigned. What's the snapshots generation?
